solana/program_v2/programs/gateway_v2/src/instructions/network/create_gatekeeper.rs

```diff
@@ -0,0 +1,91 @@
+use crate::constants::GATEKEEPER_SEED;
+use crate::errors::GatekeeperErrors;
+use crate::state::gatekeeper::{
+    Gatekeeper, GatekeeperAuthKey, GatekeeperFees, GatekeeperKeyFlags, GatekeeperState,
+};
+use crate::state::GatekeeperNetwork;
+use anchor_lang::prelude::*;
+
+pub fn create_gatekeeper(
+    ctx: Context<CreateGatekeeperAccount>,
+    data: CreateGatekeeperData,
+) -> Result<()> {
+    let authority = &mut ctx.accounts.authority;
+    let bump = *ctx.bumps.get("gatekeeper").unwrap();
+    let gatekeeper = &mut ctx.accounts.gatekeeper;
+    let network = &mut ctx.accounts.network;
+
+    if data.auth_keys.is_empty() {
+        return Err(error!(GatekeeperErrors::NoAuthKeys));
+    }
+    // Checks if there are enough auth keys to create the gatekeeper, should maybe check in NetworkKeyFlags
+    if data
+        .auth_keys
+        .iter()
+        .filter(|key| {
+            GatekeeperKeyFlags::from_bits_truncate(key.flags).contains(GatekeeperKeyFlags::AUTH)
+        })
+        .count()
+        < data.auth_threshold as usize
+    {
+        return Err(error!(GatekeeperErrors::InsufficientAuthKeys));
+    }
+
+    gatekeeper.authority = *authority.key;
+    gatekeeper.gatekeeper_bump = bump;
+    gatekeeper.gatekeeper_network = data.gatekeeper_network;
+    gatekeeper.staking_account = data.staking_account;
+    gatekeeper.token_fees = data.token_fees;
+    gatekeeper.auth_threshold = data.auth_threshold;
+    gatekeeper.auth_keys = data.auth_keys;
+    gatekeeper.gatekeeper_state = GatekeeperState::Active;
+
+    network.gatekeepers.push(gatekeeper.key());
+
+    Ok(())
+}
+/// Data for [`CreateGatekeeper`]
+#[derive(Debug, AnchorSerialize, AnchorDeserialize)]
+pub struct CreateGatekeeperData {
+    /// The [`Gatekeeper::gatekeeper_bump`].
+    pub gatekeeper_bump: u8,
+    /// The associated network for the gatekeeper
+    pub gatekeeper_network: Pubkey,
+    // staking account for the gatekeeper
+    pub staking_account: Pubkey,
```

Comment:
I would probably also pass this as an Account (`UncheckedAccount`) until it's defined.